12-offer-encoding.md

```diff
@@ -0,0 +1,823 @@
+node, and "merchant" as the shorthand for the node of someone who is
+selling or has sold something.
+
+There are two basic payment flows supported by BOLT 12:
+
+The general user-pays-merchant flow is:
+1. A merchant publishes an *offer* ("send me money"), such as on a web page or a QR code.
+2. Every user requests a unique *invoice* over the lightning network
+   using an *invoice_request* message.
+3. The merchant replies with the *invoice*.
+4. The user makes a payment to the merchant indicated by the invoice.
+
+The merchant-pays-user flow (e.g. ATM or refund):
+1. The merchant provides a user-specific *offer* ("take my money") in a web page or QR code,
+   with an amount (for a refund, also a reference to the to-be-refunded
+   invoice).
+2. The user sends an *invoice* for the amount in the *offer* (for a
+   refund, a proof that they requested the original)
+3. The merchant makes a payment to the user indicated by the invoice.
+
+## Payment Proofs and Payer Proofs
+
+Note that the normal lightning "proof of payment" can only demonstrate that an
+invoice was paid (by showing the preimage of the `payment_hash`), not who paid
+it.  The merchant can claim an invoice was paid, and once revealed, anyone can
+claim they paid the invoice, too.[1]
+
+Providing a key in *invoice_request* allows a user to prove that they were the one
+to request the invoice.  In addition, the Merkle construction of the BOLT 12
+invoice signature allows the user to selectively reveal fields of the invoice
+in case of dispute.
+
+# Encoding
+
+Each of the forms documented here are in
+[TLV](01-messaging.md#type-length-value-format) format.
+
+The supported ASCII encoding is the human-readable prefix, followed by a
+`1`, followed by a bech32-style data string of the TLVs in order,
+optionally interspersed with `+` (for indicating additional data is to
+come).
+
+## Requirements
+
+Readers of a bolt12 string:
+- if it encounters a `+` followed zero or more whitespace characters between 
+  two bech32 characters:
+  - MUST remove the `+` and whitespace.
+
+## Rationale
+
+The use of bech32 is arbitrary, but already exists in the bitcoin
+world.  We currently omit the six-character trailing checksum: QR
+codes have their own checksums anyway, and errors don't result in loss
+of funds.
+
+The use of `+` (which is ignored) allows use over limited
+text fields like Twitter:
+
+```
+lno1xxxxxxxx+
+
+yyyyyyyyyyyy+
+
+zzzzz
+```
+
+See [format-string-test.json](bolt12/format-string-test.json).
+
+## Signature Calculation
+
+All signatures are created as per
+[BIP-340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki),
+and tagged as recommended there.  Thus we define H(`tag`,`msg`) as
+SHA256(SHA256(`tag`) || SHA256(`tag`) || `msg`), and SIG(`tag`,`msg`,`key`)
+as the signature of H(`tag`,`msg`) using `key`.
+
+Each form is signed using one or more *signature TLV elements*: TLV
+types 240 through 1000.  For these
```

Comment:
Well, we want to have space for experimentation.  Since types 230-252 are two bytes more compact, I suggest we propose experimenters start from type 1000 and count down....?